components/eamxx/src/share/grid/mesh_free_grids_manager.cpp

```diff
@@ -84,11 +87,168 @@ build_grids ()
+    Kokkos::deep_copy(area, cell_area);
+    pt_grid->set_geometry_data("area", area);
+
+    // Load lat/lon if latlon_filename param is given.
+    // Else, create lat/lon with NaNs (place holders
+    // are needed for some unit tests using this
+    // grids manager).
+    if (m_params.isParameter("latlon_filename")) {
+      load_lat_lon(pt_grid);
+    } else {
+      geo_view_type lat("lat", num_local_cols);
+      geo_view_type lon("lon", num_local_cols);
+      Kokkos::deep_copy(lat, ekat::ScalarTraits<Real>::invalid());
+      Kokkos::deep_copy(lon, ekat::ScalarTraits<Real>::invalid());
+      pt_grid->set_geometry_data("lat",lat);
+      pt_grid->set_geometry_data("lon",lon);
+   }
+
+    // Load hyam/hybm if hybrid_coefficients_filename param is given.
+    if (m_params.isParameter("vertical_coordinate_filename")) {
+      load_vertical_coordinates(pt_grid);
+    }
+
+    // Load topography from file if topography_filename
+    // param is given. Else, create topography with NaNs.
+    // (place holders are needed for some unit tests
+    // using this grids manager).
+    if (m_params.isParameter("topography_filename")) {
+      load_topography(pt_grid);
+    } else {
+      geo_view_type topo("topo",num_local_cols);
+      Kokkos::deep_copy(topo,ekat::ScalarTraits<Real>::invalid());
+      pt_grid->set_geometry_data("topo",topo);
+    }
+
     add_grid(pt_grid);
     this->alias_grid("Point Grid", "Physics");
   }
 }
 
+namespace {
+
+// Helper function for creating mirror device view used to set geometry data.
+const AbstractGrid::geo_view_type cmvc(const AbstractGrid::geo_view_h_type host_view) {
+  return Kokkos::create_mirror_view_and_copy(DefaultDevice(), host_view);
+}
+
+}
+
+void MeshFreeGridsManager::
+load_lat_lon (const nonconstgrid_ptr_type& grid) const
+{
+  const int num_local_cols = grid->get_num_local_dofs();
+
+  using geo_view_host = AbstractGrid::geo_view_type::HostMirror;
+
+  // Create host mirrors for reading in data
+  std::map<std::string,geo_view_host> host_views = {
+    { "lat", geo_view_host("lat",num_local_cols) },
+    { "lon", geo_view_host("lon",num_local_cols) }
+  };
+
+  // Store view layouts
+  std::map<std::string,FieldLayout> layouts = {
+    { "lat", grid->get_2d_scalar_layout() },
+    { "lon", grid->get_2d_scalar_layout() }
+  };
+
+  // Read lat/lon into host views
+  ekat::ParameterList lat_lon_reader_pl;
+  lat_lon_reader_pl.set("Filename",m_params.get<std::string>("latlon_filename"));
+  lat_lon_reader_pl.set<std::vector<std::string>>("Field Names",{"lat","lon"});
+
+  AtmosphereInput lat_lon_reader(m_comm, lat_lon_reader_pl);
+  lat_lon_reader.init(grid, host_views, layouts);
+  lat_lon_reader.read_variables();
+  lat_lon_reader.finalize();
+
+  // Set geometry data
+  grid->set_geometry_data("lat", cmvc(host_views["lat"]));
+  grid->set_geometry_data("lon", cmvc(host_views["lon"]));
+}
+
+void MeshFreeGridsManager::
+load_vertical_coordinates (const nonconstgrid_ptr_type& grid) const
```

Comment:
There seems to be some code duplication. I'm wondering if this code duplication is a symptom of the vertical data not naturally living in a grid object.